src/actions.rs

```diff
@@ -0,0 +1,497 @@
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+}
+
+impl From<MouseActions> for ActionSequence {
+    fn from(ma: MouseActions) -> Self {
+        ActionSequence(WDActions::ActionSequence {
+            id: ma.id,
+            actions: WDActions::ActionsType::Pointer {
+                parameters: WDActions::PointerActionParameters {
+                    pointer_type: WDActions::PointerType::Mouse,
+                },
+                actions: ma.actions.into_iter().map(|x| x.into_item()).collect(),
+            },
+        })
+    }
+}
+
+/// A sequence containing `Pointer` actions for a pen device.
+#[derive(Debug, Clone)]
+pub struct PenActions {
+    /// An identifier to distinguish this sequence from others.
+    ///
+    /// Choose a meaningful string as it may be useful for debugging.
+    id: String,
+    /// The list of actions for this sequence.
+    actions: Vec<PointerAction>,
+}
+
+impl PenActions {
+    /// Create a new `PenActions` sequence.
+    pub fn new(id: &str) -> Self {
+        Self {
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+}
+
+impl From<PenActions> for ActionSequence {
+    fn from(pa: PenActions) -> Self {
+        ActionSequence(WDActions::ActionSequence {
+            id: pa.id,
+            actions: WDActions::ActionsType::Pointer {
+                parameters: WDActions::PointerActionParameters {
+                    pointer_type: WDActions::PointerType::Pen,
+                },
+                actions: pa.actions.into_iter().map(|x| x.into_item()).collect(),
+            },
+        })
+    }
+}
+
+/// A sequence containing `Pointer` actions for a touch device.
+#[derive(Debug, Clone)]
+pub struct TouchActions {
+    /// An identifier to distinguish this sequence from others.
+    ///
+    /// Choose a meaningful string as it may be useful for debugging.
+    id: String,
+    /// The list of actions for this sequence.
+    actions: Vec<PointerAction>,
+}
+
+impl TouchActions {
+    /// Create a new `TouchActions` sequence.
+    pub fn new(id: &str) -> Self {
+        Self {
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+}
+
+impl From<TouchActions> for ActionSequence {
+    fn from(ta: TouchActions) -> Self {
+        ActionSequence(WDActions::ActionSequence {
+            id: ta.id,
+            actions: WDActions::ActionsType::Pointer {
+                parameters: WDActions::PointerActionParameters {
+                    pointer_type: WDActions::PointerType::Touch,
+                },
+                actions: ta.actions.into_iter().map(|x| x.into_item()).collect(),
+            },
+        })
+    }
+}
+
+/// A sequence of actions to be performed.
+///
+/// See the documentation for [`Actions`] for more details.
+#[derive(Debug)]
+pub struct ActionSequence(pub(crate) WDActions::ActionSequence);
+
+/// Each sequence type implements `InputSource` which provides a `pause()` and a `then()`
```

Comment:
Let's have a summary first line, something along the lines of:

> An source capable of providing inputs for a browser action chain.

And also link to the spec section on input sources.

Reply:
I don't think you ended up adding a link to the spec, unless I missed it?

Reply:
Ah, I missed the link. Added now.